File: coding-practice/LeetCode 75/Container With Most Water/Solution.py
import time

def measure_time(func):
    def wrapper(*args, **kwargs):
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()
        print(f"Time taken to run '{func.__name__} function': {(end_time - start_time) * 1000} miliseconds")
        return result
    return wrapper


class Solution:
    def __init__(self) -> None:
        pass

    @measure_time
    def maxArea(self, height: list):
        left = 0
        right = len(height) - 1
        max_area = 0

        while left < right:
            area = min(height[left], height[right]) * (right - left)

            max_area = max(max_area, area)

            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        
        print(f'max area : {max_area}')

height = [1,8,6,2,5,4,8,3,7]
solution = Solution()
print(solution.maxArea(height))

File: coding-practice/LeetCode 75/Container With Most Water/C++/Solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution{
    private:

    public:
    int max_element(vector<int>& height){
        int maximum = height[0]; 
        for(int i = 1; i < height.size(); i++){
            if(maximum < height[i]){
                maximum = height[i];
            }
        }

        return maximum;
    }
    
    int min_element(vector<int>& height){
        int minimum = height[0]; 
        for(int i = 1; i < height.size(); i++){
            if(minimum > height[i]){
                minimum = height[i];
            }
        }

        return minimum;
	}
	
	int min(int x, int y){
		if(x > y){
			return y;
		}
		return x;
	}
    
    
    int maxArea(vector<int>& height){
    	int left, right, max_area;
    	left = 0;
    	right = height.size() - 1;
    	max_area = 0;
    	
    	while(left < right){
    		int area = min(height[left], height[right]) * (right - left);
    		max_area = max(max_area, area);
    		
    		if(height[left] < height[right]){
    			left += 1;
			}else{
				right -= 1;
			}
		}
		return max_area;
	}
};


int main(){
    Solution obj;

    int arr[] = {1,8,6,2,5,4,8,3,7};
	vector<int> height(arr, arr + sizeof(arr) / sizeof(int));
    
	int maxArea = obj.maxArea(height);
	cout << "The max area is: " << maxArea;
    return 0;
}
